### ISSUE TITLE
train errors

### ISSUE THREAD

**Issue by KangolHsu** [EunhyeokPark/caffe-WQ#1]
when i train my model with “WeightQuantConvolution”，error emerged：
`F1020 18:35:05.314141 49423 layer_factory.hpp:81] Check failed: registry.count(type) == 1 (0 vs. 1) Unknown layer type: WeightQuantConvolution (known types: AbsVal, Accuracy, ArgMax, BNLL, BatchNorm, BatchReindex, Bias, CenterLoss, Concat, ContrastiveLoss, Convolution, Crop, Data, Deconvolution, Dropout, DummyData, ELU, Eltwise, Embed, EuclideanLoss, Exp, Filter, Flatten, HDF5Data, HDF5Output, HingeLoss, Im2col, ImageData, InfogainLoss, InnerDistance, InnerProduct, Input, LRN, LSTM, LSTMUnit, Log, MVN, MemoryData, MultinomialLogisticLoss, Normalize, PReLU, Parameter, Pooling, Power, Python, RNN, ReLU, Reduction, Reshape, SPP, Scale, Sigmoid, SigmoidCrossEntropyLoss, Silence, Slice, Softmax, SoftmaxWithLoss, Split, TanH, Threshold, Tile, WeightLogQuantReLU, WeightQuantInnerProduct, WindowData)
`
maybe your caffe.proto may miss something？

**Comment by KangolHsu** [comment]
I set CUDNN=1 works fine
but new error ocurrs (with DEBUG=1):

`*** Aborted at 1508753023 (unix time) try "date -d @1508753023" if you are using GNU date ***

PC: @     0x7f79c70df9e3 caffe::QuantInfo<>::wp()

*** SIGSEGV (@0x1050d9ffffc) received by PID 55438 (TID 0x7f79c8fa7ac0) from PID 228589564; stack trace: ***
    @     0x7f79c51c84b0 (unknown)
    @     0x7f79c70df9e3 caffe::QuantInfo<>::wp()
    @     0x7f79c70defdb caffe::QuantInfo<>::wEnt()
    @     0x7f79c70ddeae caffe::eh_weight_quant_update<>()
    @     0x7f79c70bb40b caffe::WeightQuantConvolutionLayer<>::Forward_gpu()
    @           0x4242ca caffe::Layer<>::Forward()
    @     0x7f79c6ef6153 caffe::Net<>::ForwardFromTo()
    @     0x7f79c6ef5dbb caffe::Net<>::Forward()
    @     0x7f79c6f26aa0 caffe::Solver<>::Test()
    @     0x7f79c6f267ac caffe::Solver<>::TestAll()
    @     0x7f79c6f24fe2 caffe::Solver<>::Step()
    @     0x7f79c6f24ad3 caffe::Solver<>::Solve()
    @           0x41ecb6 train()
    @           0x421025 main
    @     0x7f79c51b3830 __libc_start_main
    @           0x41d5c9 _start
    @                0x0 (unknown)
Segmentation fault (core dumped)
`

**Comment by cmalab** [comment]
My code has been tested in the following environments: Ubuntu 16.04, cuDNNv5, NVIDIA375.66
Please let me know your detailed experiment system spec. Furthermore, I highly recommend you to start with the following script: https://github.com/EunhyeokPark/script_for_WQ

The modified layers of quantization code require several additional parameters. Above script shows how to set up the parameters in detail.

**Comment by EunhyeokPark** [comment]
My code has been tested in the following environments: Ubuntu 16.04, cuDNNv5, NVIDIA375.66
Please let me know your detailed experiment system spec. Furthermore, I highly recommend you to start with the following script: https://github.com/EunhyeokPark/script_for_WQ

The modified layers of quantization code require several additional parameters. Above script shows how to set up the parameters in detail.

**Comment by KangolHsu** [comment]
@EunhyeokPark  
I had used the script quantization_alexnet.py ，it can works fine，
My environments：

```
NVIDIA-SMI 375.66                 Driver Version: 375.66 
Ubuntu 16.04
cudnn-8.0-linux-x64-v5.0-ga
```
 
Some converted prototxt

```
layer {
  name: "conv1"
  type: "WeightQuantConvolution"
  bottom: "data"
  top: "conv1"
  param {
    lr_mult: 1.0
    decay_mult: 1.0
  }
  param {
    lr_mult: 2.0
    decay_mult: 0.0
  }
  param {
    lr_mult: 0.0
    decay_mult: 0.0
  }
  param {
    lr_mult: 0.0
    decay_mult: 0.0
  }
  param {
    lr_mult: 0.0
    decay_mult: 0.0
  }
  convolution_param {
    num_output: 96
    pad: 1
    kernel_size: 3
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}
layer {
  name: "slice1"
  type: "Slice"
  bottom: "conv1"
  top: "slice1_1"
  top: "slice1_2"
  slice_param {
    slice_dim: 1
  }
}
layer {
  name: "etlwise1"
  type: "Eltwise"
  bottom: "slice1_1"
  bottom: "slice1_2"
  top: "eltwise1"
  eltwise_param {
    operation: MAX
  }
}
layer {
  name: "pool1"
  type: "Pooling"
  bottom: "eltwise1"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}

```

I had added some layers ，i will check it carefully。

**Comment by KangolHsu** [comment]
Maybe i used the Eltwise layer which did not  reimplemented as ReLU （WeightLogQuantReLU）.

**Comment by hewumars** [comment]
@KangolHsu 请问下WQ量化，您有重现出来吗？

**Comment by KangolHsu** [comment]
@hewumars  还没有

**Comment by hewumars** [comment]
@KangolHsu 觉得这种方式还是有点复杂，工程化比较困难，你这边有了解过一些比较好的，不需要训练的量化方法吗？比如tensorrt那种思路的。

**Comment by KangolHsu** [comment]
@hewumars 貌似都得finetune一下，不然直接量化精度损失较大

**Comment by hewumars** [comment]
@KangolHsu 我试过tensorrt它的损失控制的很好，基本在1%以内，可惜不开源。还是要找找。。。暂时没有发现直接量化的方法。

**Comment by KangolHsu** [comment]
@hewumars TensorRT只能在NVIDIA TX上跑？是的话那就没啥意义了。

**Comment by EunhyeokPark** [comment]
I don't know Chinese, so I barely understand the mention using google translator :)

Basically, this method is designed to automatically find the optimal quantization level without human intervention. However, not all DNN layers are supported yet, and there are some difficulties of setting several hyper-parameters internally in Caffe. I provide an example of automated code for this, so please check:
https://github.com/EunhyeokPark/script_for_WQ

**Comment by EunhyeokPark** [comment]
In the case of TensorRT, as far as I know, Pascal based GPUs support 8-bit operation.

**Comment by hewumars** [comment]
Thanks,If I have questions,I'll ask you again.  I am researching your train code.The quanModel has been generated with scripe_for_WQ.I found it just added 3 blobs with blob[2]=[0,weight_level,1]，blob[3] = quant_info，blob[4] = blob[0].

**Comment by hewumars** [comment]
@KangolHsu PC: @ 0x7f79c70df9e3 caffe::QuantInfo<>::wp() 你这个地方报错，解决了吗？我也遇到了。
